ensime_shared/ensime.py

```diff
@@ -136,16 +136,23 @@ def queue_poll(self, sleep_t=0.5):
 
         Value of sleep is low to improve responsiveness.
         """
+        connection_alive = True
         while self.running:
             if self.ws:
                 def logger_and_close(m):
                     self.log("Websocket exception: {}".format(m))
-                    self.teardown()
+                    if not self.running:
```

Comment:
Yes, it can be hit. It's hit every time there's a Websocket exception and we haven't called `teardown`, that is, when an unexpected websocket exception happens and the user hasn't exited from vim. In that case, `self.running` is still `True` and we need to close the editor (we choose to). Otherwise we mark the connection as dead and exit without executing `time.sleep`.